File: AWS/iac-microbenchmark/lambdas/typescript/rsa3072_decrypt.ts

import { KMSClient, DecryptCommand } from "@aws-sdk/client-kms";
import * as crypto from "crypto";
import { Context, APIGatewayProxyResult, APIGatewayEvent } from 'aws-lambda';

// Initialize KMS Client
const kmsClient = new KMSClient({});

export const handler = async (event: APIGatewayEvent, context: Context): Promise<APIGatewayProxyResult> => {

  // Get the KMS key ID from environment variables
  const rsaKmsKeyId = process.env.RSA3072_KMS_KEY_ID;

  if (!rsaKmsKeyId) {
    return {
      statusCode: 400,
      body: JSON.stringify({ message: "KMS Key ID not provided" }),
    };
  }

  // Get encrypted data from the event
  const encryptedAesKey = Buffer.from(event.body ? JSON.parse(event.body).encrypted_aes_key : '', 'base64');
  const iv = Buffer.from(event.body ? JSON.parse(event.body).iv : '', 'base64');
  const ciphertext = Buffer.from(event.body ? JSON.parse(event.body).ciphertext : '', 'base64');

  if (!encryptedAesKey || !iv || !ciphertext) {
    return {
      statusCode: 400,
      body: JSON.stringify({ message: "Missing encrypted data in the event" }),
    };
  }

  try {
    // Decode base64 values

    // Decrypt the AES key using KMS
    const decryptCommand = new DecryptCommand({
      CiphertextBlob: encryptedAesKey,
      KeyId: rsaKmsKeyId,
      EncryptionAlgorithm: "RSAES_OAEP_SHA_256",
    });

    const kmsResponse = await kmsClient.send(decryptCommand);
    const aesKey = kmsResponse.Plaintext;

    if (!aesKey) {
      return {
        statusCode: 500,
        body: JSON.stringify({ message: "Failed to decrypt AES key" }),
      };
    }

    // Decrypt the ciphertext using AES-CTR
    const decipher = crypto.createDecipheriv('aes-256-ctr', aesKey as Buffer, iv);
    const decrypted = Buffer.concat([decipher.update(ciphertext), decipher.final()]); // Return the plaintext data
    return {
      headers: {
        "Access-Control-Allow-Origin": "*",
        "content-type": "application/json"
      },
      statusCode: 200,
      body: JSON.stringify({ plaintext: decrypted.toString('utf-8') }),
    };
  } catch (error) {
    console.error("Error decrypting data: ", error);
    return {
      headers: {
        "Access-Control-Allow-Origin": "*",
        "content-type": "application/json"
      },
      statusCode: 500,
      body: JSON.stringify({ message: "Decryption failed" }),
    };
  }
};

File: AWS/iac-microbenchmark/ec2/python/rsa3072_decrypt.py
import base64
import json
import os
import sys
import boto3
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

# AWS KMS Client
kms_client = boto3.client('kms')

def main():
    # Get the name argument from sys.argv
    request_json_raw = sys.argv[1]

    request_json = json.loads(request_json_raw)
    # Get the KMS key ID from environment variables
    rsa_kms_key_id = os.environ['RSA3072_KMS_KEY_ARN']
    
    # Get the data from the input payload
    encrypted_aes_key_b64 = request_json.get('encrypted_aes_key')
    iv_b64 = request_json.get('iv')
    ciphertext_b64 = request_json.get('ciphertext')

    # Decode base64 values
    encrypted_aes_key = base64.b64decode(encrypted_aes_key_b64)
    iv = base64.b64decode(iv_b64)
    ciphertext = base64.b64decode(ciphertext_b64)

    # Decrypt the AES key using KMS
    response = kms_client.decrypt(
        CiphertextBlob=encrypted_aes_key,
        KeyId=rsa_kms_key_id,
        EncryptionAlgorithm="RSAES_OAEP_SHA_256"
    )
    aes_key = response['Plaintext']

    # Decrypt the data using AES
    cipher = Cipher(algorithms.AES(aes_key), modes.CTR(iv), backend=default_backend())
    decryptor = cipher.decryptor()
    plaintext = decryptor.update(ciphertext) + decryptor.finalize()

    # Return the plaintext as a JSON response
    result = {
        'plaintext': plaintext.decode('utf-8')
    }

    print(json.dumps(result))

if __name__ == "__main__":
    main()
